gef.py

```diff
@@ -6760,55 +6749,119 @@ def do_invoke(self, argv):
             info("No Tcache in this version of libc")
             return
 
-        arena = GlibcArena("*{:s}".format(argv[0])) if len(argv) == 1 else get_main_arena()
-
-        if arena is None:
-            err("Invalid Glibc arena")
-            return
-
-        # Get tcache_perthread_struct for this arena
-        heap_base = HeapBaseFunction.heap_base()
-        if heap_base is None:
-            err("No heap section")
+        current_thread = gdb.selected_thread()
+        if current_thread is None:
+            err("Couldn't find current thread")
             return
-        addr = heap_base + 0x10
 
-        gef_print(titlify("Tcachebins for arena {:#x}".format(int(arena))))
-        for i in range(GlibcArena.TCACHE_MAX_BINS):
-            if get_libc_version() < (2, 30):
-                count = ord(read_memory(addr + i, 1))
+        threads = gdb.selected_inferior().threads()
+        if argv:
+            if "all" in argv:
+                tids = [t.num for t in threads]
             else:
-                count = u16(read_memory(addr + 2 * i, 2))
-            chunk = arena.tcachebin(i)
-            chunks = set()
-            m = []
+                tids = self.check_thread_ids(argv)
+        else:
+            tids = [current_thread.num]
 
-            # Only print the entry if there are valid chunks. Don't trust count
-            while True:
-                if chunk is None:
-                    break
+        # As a nicety, we want to display threads in ascending order by gdb number
+        for thread in sorted(threads, key=lambda t: t.num):
+            if thread.num not in tids:
+                continue
 
-                try:
-                    m.append("{:s} {:s} ".format(LEFT_ARROW, str(chunk)))
-                    if chunk.address in chunks:
-                        m.append("{:s} [loop detected]".format(RIGHT_ARROW))
+            thread.switch()
+            tcache_addr = gdb.parse_and_eval("(void *) tcache")
```

Comment:
Where is this global defined? Can you explain a bit more?
It doesn't seem to exist on non multi-threaded binaries, and make `heap bins tcache` throw an exception

Tested with
```c
void main()
{
char *a = malloc(0x20);
char *b = malloc(0x20);
char *c = malloc(0x20);
free(b);
__asm__("int3");
}
```

Result:
```
0:000 ➤  run
[...]
0:000 ➤  heap bins tcache

─────────────────────────────── Exception raised ─────────────────────────────── 
error: Cannot find thread-local storage for process 468212, shared library /lib/x86_64-linux-gnu/libc.so.6:
Cannot find thread-local variables on this target
───────────────────────────── Detailed stacktrace ────────────────────────────── 
↳ File "/tmp/tmp.Gy2Zp7Xfjj/gef/gef.py", line 6772, in do_invoke()
    →     tcache_addr = gdb.parse_and_eval("(void *) tcache")                         
[...]
```

Reply:
`tcache` is a thread local global defined at https://elixir.bootlin.com/glibc/glibc-2.33/source/malloc/malloc.c#L3078

I'm confused as to why it can't be found without linking `lpthread` - looks like this is a gap in my understanding of how thread local storage works?

Be that as it may, it seems like we can fix this be falling back to the strategy that gef used before this PR - `tcache` is the first allocation, so it's at a predictable spot in the main arena each time. Implemented in e3cf5a2e1e2475300a9a11cf718118eedb2d8b05